src/catalog/error.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fmt;
+
+use backtrace::Backtrace;
+
+#[derive(Debug)]
+pub struct Error {
+    kind: ErrorKind,
+    backtrace: Backtrace,
+}
+
+#[derive(Debug)]
+pub(crate) enum ErrorKind {
+    Corruption { detail: String },
+    IdExhaustion,
+    UnknownDatabase(String),
+    UnknownSchema(String),
+    UnknownItem(String),
+    DatabaseAlreadyExists(String),
```

Comment:
Not necessarily now, but if we factored the item type out like so:

```rust
pub struct Error {
    kind: ErrorKind
    item: Box<str>, // or String whatever
    backtrace: Backtrace,
}

pub(crate) enum ErrorKind {
    Corruption,
    IdExhaustion,
    UnknownDatabase,
    //..snip..
    Storage(rusqlite::Error),
}
```

it would be possible to do the same factoring that I have in the other pr and
do something like:

```rust
impl Transaction {
    pub fn load_database_id(&self, database_name: &str) -> Result<i64, Error> {
        self.load_id(
            database_name,
            "SELECT id FROM databases WHERE name = ?",
            params![database_name],
            ErrorKind::UnknownDatabase,
        )
    }

    // snip

    pub fn load_id(&self, item: &str, query: &str, params: &[&dyn ToSql]) -> Result<i64, Error {
        match self
            .inner
            .prepare_cached(query).sql_ctx("prepare_cached", item)?
            .query_row(params, |row| row.get(0))
        {
            Ok(id) => Ok(id),
            Err(rusqlite::Error::QueryReturnedNoRows) => Err(Error::new(
                item.to_owned(), ErrorKind::UnknownDatabase,
            )),
            Err(err) => Err(Error::sqlite("loading id", item)), // providing context
        }
    }
}
```

Which will just make sure that we have reliable errors as we add more dsl
methods.

What do you think?